### PR TITLE
#5509 - Thread Upvotes Drawer

### PR DESCRIPTION
**Pull request by Miaplacidus**
This PR adds a 'View Upvotes' trigger on the discussions page and the view thread page that, when clicked, opens a drawer with details on the upvotes the current thread has received. This PR also adds a 'View Upvotes' trigger on individual thread comments. 

## Link to Issue
Closes: #5509, #6472

## Description of Changes
- creates a drawer for viewing upvotes that contains the content of the thread/comment and a table with the name of the upvoter, the vote weight, and the time the vote was made

## Test Plan
- visit the discussions page and click 'View Upvotes'
- visit the page of any thread with upvotes and click 'View Upvotes'

## Other Considerations
- there is a bug wherein clicking outside of the drawer does not collapse it. That will be handled in ticket #6432  
- the timestamp data for an upvote is not available on the discussions page. That will be handled in ticket #6459

### CODE REVIEW

**Comment by Israellund** [comment]
This looks great! I just had one possible suggestion/possible change. If there are a lot of upvotes and the user scrolls all the way down, the only way to close the drawer is to scroll all the way back up. I know this wont be a problem once the user is able to click outside of the drawer and have it close, but what if we had the arrows fixed with a high z-index so that it stays above the content and fixed in the upper left corner? 
@Miaplacidus @masvelio @mzparacha @lzach83 Anyone have any thoughts for or against this? 

https://github.com/hicommonwealth/commonwealth/assets/69872984/8e847277-dfad-474f-b44f-b8e3668dfc49

**Review by Israellund — COMMENTED** [review]
LGTM
Great job!

**Review by Israellund — APPROVED** [review]

**Comment by Miaplacidus** [comment]
> 1. We should adjust `View upvotes` to be like a button, similar to others in the row. So with hover effect and padding to left and right
> 
>  Kapture.2024-01-25.at.12.57.36.mp4 
> 2. On mobile I can't see last row. I assume this is because drawer has height of 100vh, but the top navbar should be subtracted
> 
>  Kapture.2024-01-25.at.13.00.55.mp4 
> 3. in comments drawer the list of upvoters is not refresh after adding/removing upvote. it works fine on thread drawer
> 
>  Kapture.2024-01-25.at.13.06.06.mp4 
> Great job! Generally looks and feels good. Found some small issues and have some requests for changing a code. The biggest problem here is #6432 though

@masvelio Fixed issues 1 and 2, but 3 is a nightmare! Will come back to that later!

File: packages/commonwealth/client/scripts/views/components/view_comment_upvotes_drawer.tsx [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import { useFetchProfilesByAddressesQuery } from 'client/scripts/state/api/profiles';
+import React from 'react';
+import app from 'state';
+import { ViewUpvotesDrawer } from './view_upvotes_drawer';
+
+type ViewCommentUpvotesDrawerProps = {
+  comment?: Comment<any>;
+};
+
+export const ViewCommentUpvotesDrawer = ({
```
[the remark this answers is not in the document]
Comment (Miaplacidus) [reply packages/commonwealth/client/scripts/views/components/view_comment_upvotes_drawer.tsx]: Done!

File: packages/commonwealth/client/scripts/views/components/view_thread_upvotes_drawer.tsx [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import { useFetchProfilesByAddressesQuery } from 'client/scripts/state/api/profiles';
+import React from 'react';
+import app from 'state';
+import { ViewUpvotesDrawer } from './view_upvotes_drawer';
+
+type ViewThreadUpvotesDrawerProps = {
+  thread?: Thread;
+};
+
+export const ViewThreadUpvotesDrawer = ({
```
[the remark this answers is not in the document]
Comment (Miaplacidus) [reply packages/commonwealth/client/scripts/views/components/view_thread_upvotes_drawer.tsx]: They are quite similar. Initially, I had them in the same component, but I found that separating the drawers into two separate components was easier to comprehend than having logic to handle the different attributes on reactions.

File: packages/commonwealth/client/scripts/views/components/view_upvotes_drawer.tsx [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+export const ViewUpvotesDrawer = ({
+  header,
+  reactorData,
+  contentBody,
+  author,
+  publishDate,
+}: ViewUpvotesDrawerProps) => {
+  const [isUpvoteDrawerOpen, setIsUpvoteDrawerOpen] = useState(false);
+
+  const getColumnInfo = () => {
```
[the remark this answers is not in the document]
Comment (Miaplacidus) [reply packages/commonwealth/client/scripts/views/components/view_upvotes_drawer.tsx]: Good point. Done!

File: packages/commonwealth/client/scripts/views/components/view_upvotes_drawer.tsx [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+              authorChainId={getAuthorCommunityId(author)}
+              publishDate={publishDate}
+              showUserAddressWithInfo={false}
+            />
+          </div>
+          <div className="upvoted-content-body">
+            <QuillRenderer doc={contentBody} cutoffLines={10} />
+          </div>
+        </div>
+        {reactorData && (
```
[the remark this answers is not in the document]
Comment (Miaplacidus) [reply packages/commonwealth/client/scripts/views/components/view_upvotes_drawer.tsx]: I've fixed this issue. Now, when there is no data, the user sees placeholder text telling them that there are no upvotes to view.

File: packages/commonwealth/client/scripts/views/components/view_upvotes_drawer.tsx [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+  const getAuthorCommunityId = (contentAuthor) => {
+    if (contentAuthor instanceof MinimumProfile) {
+      return contentAuthor?.chain;
+    } else if (contentAuthor instanceof Account) {
+      return contentAuthor.community.id;
+    }
+  };
+
+  return (
+    <>
```
[the remark this answers is not in the document]
Comment (Miaplacidus) [reply packages/commonwealth/client/scripts/views/components/view_upvotes_drawer.tsx]: Nice catch! I've pulled upvotes drawer styling into its own file.

File: packages/commonwealth/client/scripts/views/components/view_upvotes_drawer.tsx [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        key: 'timestamp',
+        header: 'Timestamp',
+        numeric: true,
+        sortable: true,
+        chronological: true,
+      },
+    ];
+  };
+
+  const voterRow = (voter) => {
```
[the remark this answers is not in the document]
Comment (Miaplacidus) [reply packages/commonwealth/client/scripts/views/components/view_upvotes_drawer.tsx]: Done!

File: packages/commonwealth/client/scripts/views/pages/discussions/ThreadCard/ThreadOptions/ThreadOptions.tsx [hunk 7]
```diff
@@ -83,54 +85,58 @@ export const ThreadOptions = ({
           e.preventDefault();
```
[the remark this answers is not in the document]
Comment (Miaplacidus) [reply packages/commonwealth/client/scripts/views/pages/discussions/ThreadCard/ThreadOptions/ThreadOptions.tsx:85]: Thanks for the catch, @masvelio! I'll keep this in mind when working on #6432!

File: packages/commonwealth/client/scripts/views/components/UpvoteDrawer/ViewThreadUpvotesDrawer.tsx [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    const reactor = reactors.find((r) => r.address === profile.address);
+
+    return {
+      name: profile.name,
+      avatarUrl: profile.avatarUrl,
+      address: profile.address,
+      updated_at: reactor?.updated_at,
+      voting_weight: reactor?.voting_weight,
+    };
+  });
```
Comment (masvelio) [inline packages/commonwealth/client/scripts/views/components/UpvoteDrawer/ViewThreadUpvotesDrawer.tsx:33]: let's move this code to useUpvoteDrawer hook and use it here and in ViewCommentUpvotesDrawer

**Review by masvelio — APPROVED** [review]
Found also an issue when I react in discussion page, then I can see error in the console about NaN value; nan value is displaued in the summary row; no value is displayed in the weight column

https://github.com/hicommonwealth/commonwealth/assets/14819225/24f5e72b-10a1-4338-be12-b7341b219f31

Not sure what is the plan for #3 from previous code review? In case you want to merge this PR, I am approving it, but please track those issues in other ticket + resolve the comments (use custom hook also)

**Comment by Miaplacidus** [comment]
> Found also an issue when I react in discussion page, then I can see error in the console about NaN value; nan value is displaued in the summary row; no value is displayed in the weight column
> 
>  Kapture.2024-01-26.at.13.16.34.mp4 
> Not sure what is the plan for #3 from previous code review? In case you want to merge this PR, I am approving it, but please track those issues in other ticket + resolve the comments (use custom hook also)

Hey, @masvelio! I noticed the issue on the discussions page and have pushed up a fix for it!

**Comment by Miaplacidus** [comment]
Added ticket #6494 to handle bug wherein upvoting a comment doesn't immediately appear in the upvotes drawer.

**Pull request merged**